Assets/Scripts/Systems/Modules/ModulesInfo.cs

```diff
@@ -7,97 +7,169 @@
+                }
+
+                foreach (ModuleSlot moduleSlot in _moduleSlots)
+                {
+                    AddModuleSlot(moduleSlot);
+                }
             }
+        }
 
-            if (Modules == null)
+        public Module GetModule(Coordinates modulePos)
+        {
+            return _grid[modulePos.y, modulePos.x];
+        }
+
+        public bool AddModule(Module module)
+        {
+            if (ModulePositionValid(module))
             {
-                Modules = new List<Module>();
+                Coordinates rootPos = module.RootPosition;
+                foreach (Coordinates coords in module.Data.GridPositions)
+                {
+                    if (rootPos.x + coords.x >= 0 && rootPos.x + coords.x < _columnLength &&
+                        rootPos.y + coords.y >= 0 && rootPos.y + coords.y < _rowHeight)
+                    {
+                        _grid[rootPos.y + coords.y, rootPos.x + coords.x] = module;
+                    }
+                }
+
+                _modules.Add(module);
+                return true;
             }
+
+            return false;
         }
 
-        public void AddModule(Module module, int row, int column)
+        public void RemoveModule(Coordinates modulePos)
         {
-            foreach(Coordinates coords in module.Data.GridPositions)
+            Module moduleToRemove = _grid[modulePos.y, modulePos.x];
+            Coordinates rootPos = moduleToRemove.RootPosition;
+            
+            foreach (Coordinates coords in moduleToRemove.Data.GridPositions)
             {
-                if (column + coords.x >= 0 && column + coords.x < ColumnLength &&
-                    row + coords.y >= 0 && row + coords.y < RowHeight)
+                if (rootPos.x + coords.x > 0 && rootPos.x + coords.x < _columnLength &&
+                    rootPos.y + coords.y > 0 && rootPos.y + coords.y < _rowHeight)
                 {
-                    Grid[row + coords.y, column + coords.x] = module;
+                    if (_grid[rootPos.y + coords.y, rootPos.x + coords.x] == moduleToRemove)
+                    {
+                        _grid[rootPos.y + coords.y, rootPos.x + coords.x] = null;
+                    }
                 }
             }
 
-            if (!Modules.Contains(module))
+            _modules.Remove(moduleToRemove);
+        }
+
+        public bool AddModuleSlot(ModuleSlot moduleSlot)
+        {
+            Coordinates pos = moduleSlot.Position;
+            if (pos.x > 0 && pos.x < _columnLength &&
+                pos.y > 0 && pos.y < _rowHeight &&
+                _slotGrid[pos.y, pos.x] == null)
             {
-                Modules.Add(module);
+                _slotGrid[pos.y, pos.x] = moduleSlot;
+                return true;
             }
 
-            module.PivotPosition = new Coordinates
-            {
-                x = column,
-                y = row
-            };
+            return false;
         }
```

Comment:
Don't think you should need this function, since you can't currently change the shape of the module grid at runtime, so being able to add module slots doesn't really make sense